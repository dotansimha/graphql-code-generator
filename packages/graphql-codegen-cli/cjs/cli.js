module.exports = require('esm')(module, { force: true })('../dist/cli.js');
