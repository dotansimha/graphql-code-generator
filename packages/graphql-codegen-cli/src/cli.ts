import { generate } from './generate-and-save';
import { init } from './init';
import { createContext } from './config';
import { lifecycleHooks } from './hooks';
import { DetailedError } from '@graphql-codegen/plugin-helpers';

export function runCli(cmd: string): Promise<any> {
  ensureGraphQlPackage();

  switch (cmd) {
    case 'init':
      return init();
    default: {
      return createContext().then(context => {
        return generate(context).catch(async error => {
          await lifecycleHooks(context.getConfig().hooks).onError(error.toString());

          throw error;
        });
      });
    }
  }
}

export function ensureGraphQlPackage() {
  try {
    require('graphql');
  } catch (e) {
    throw new DetailedError(
      `Unable to load "graphql" package. Please make sure to install "graphql" as a dependency!`,
      `
  To install "graphql", run:
    yarn add graphql
  Or, with NPM:
    npm install --save graphql
`
    );
  }
}
