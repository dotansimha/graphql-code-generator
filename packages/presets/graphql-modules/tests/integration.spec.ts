import { useMonorepo } from '@graphql-codegen/testing';
import { executeCodegen } from '@graphql-codegen/cli';
import { normalize } from 'path';

const monorepo = useMonorepo({
  dirname: __dirname,
});

const options = {
  generates: {
    './tests/test-files/modules': {
      schema: './tests/test-files/modules/*/types/*.graphql',
      plugins: ['typescript', 'typescript-resolvers'],
      preset: 'graphql-modules',
      presetConfig: {
        baseTypesPath: 'global-types.ts',
        filename: 'module-types.ts',
        encapsulateModuleTypes: 'none',
      },
    },
  },
};

describe('Integration', () => {
  monorepo.correctCWD();

  beforeEach(() => {
    jest.useFakeTimers();
  });

  // In this test, we make sure executeCodegen passes on a list of Sources as an extension
  // This is very important
  test('should generate a base output and 4 for modules', async () => {
    try {
      const output = await executeCodegen(options);

      expect(output.length).toBe(5);
      expect(normalize(output[0].filename)).toMatch(normalize(`/modules/global-types.ts`));
      expect(normalize(output[1].filename)).toMatch(normalize(`/modules/blog/module-types.ts`));
      expect(normalize(output[2].filename)).toMatch(normalize(`/modules/common/module-types.ts`));
      expect(normalize(output[3].filename)).toMatch(normalize(`/modules/dotanions/module-types.ts`));
      expect(normalize(output[4].filename)).toMatch(normalize(`/modules/users/module-types.ts`));
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(e);
      throw e;
    }
  });

  test.only('should allow to override importBaseTypesFrom correctly', async () => {
    const output = await executeCodegen({
      generates: {
        './tests/test-files/modules': {
          schema: './tests/test-files/modules/*/types/*.graphql',
          plugins: ['typescript', 'typescript-resolvers'],
          preset: 'graphql-modules',
          presetConfig: {
            importBaseTypesFrom: '@types',
            baseTypesPath: 'global-types.ts',
            filename: 'module-types.ts',
            encapsulateModuleTypes: 'none',
          },
        },
      },
    });
    const importStatement = `import * as Types from "@types";`;

    expect(output.length).toBe(5);
    expect(output[1].content).toMatch(importStatement);
    expect(output[2].content).toMatch(importStatement);
    expect(output[3].content).toMatch(importStatement);
    expect(output[4].content).toMatch(importStatement);
  });

  test('each module-types should include a relative import to glob-types module', async () => {
    const output = await executeCodegen(options);
    const importStatement = `import * as Types from "../global-types";`;

    expect(output.length).toBe(5);
    expect(output[1].content).toMatch(importStatement);
    expect(output[2].content).toMatch(importStatement);
    expect(output[3].content).toMatch(importStatement);
    expect(output[4].content).toMatch(importStatement);
  });

  test('each module-types should export Resolvers', async () => {
    const output = await executeCodegen(options);
    const exportStatemment = `export interface Resolvers `;

    expect(output.length).toBe(5);
    expect(output[1].content).toMatch(exportStatemment);
    expect(output[2].content).toMatch(exportStatemment);
    expect(output[3].content).toMatch(exportStatemment);
    expect(output[4].content).toMatch(exportStatemment);
  });

  test('dotanions module should export DefinedFields, Schema Types with Picks and resolvers', async () => {
    const output = await executeCodegen(options);

    expect(output.length).toBe(5);
    expect(output[3].content).toMatchSnapshot();
  });
});
