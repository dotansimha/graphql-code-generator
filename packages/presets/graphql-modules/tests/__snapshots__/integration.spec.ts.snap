// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration dotanions module should export DefinedFields, Schema Types with Picks and resolvers 1`] = `
"import * as Types from \\"../global-types\\";
import * as gm from \\"graphql-modules\\";
interface DefinedFields {
  Paypal: 'id' | 'url';
  CreditCard: 'id' | 'cardNumber' | 'cardOwner';
  Donation: 'id' | 'sender' | 'recipient' | 'amount';
  Mutation: 'donate';
  User: 'paymentOptions';
};

interface DefinedInputFields {
  DonationInput: 'user' | 'amount' | 'paymentOption';
};

export type Paypal = Pick<Types.Paypal, DefinedFields['Paypal']>;
export type CreditCard = Pick<Types.CreditCard, DefinedFields['CreditCard']>;
export type PaymentOption = Types.PaymentOption;
export type User = Types.User;
export type Donation = Pick<Types.Donation, DefinedFields['Donation']>;
export type DonationInput = Pick<Types.DonationInput, DefinedInputFields['DonationInput']>;
export type Mutation = Pick<Types.Mutation, DefinedFields['Mutation']>;

export type PaypalResolvers = Pick<Types.PaypalResolvers, DefinedFields['Paypal'] | '__isTypeOf'>;
export type CreditCardResolvers = Pick<Types.CreditCardResolvers, DefinedFields['CreditCard'] | '__isTypeOf'>;
export type DonationResolvers = Pick<Types.DonationResolvers, DefinedFields['Donation'] | '__isTypeOf'>;
export type MutationResolvers = Pick<Types.MutationResolvers, DefinedFields['Mutation']>;
export type UserResolvers = Pick<Types.UserResolvers, DefinedFields['User']>;

export interface Resolvers {
  Paypal?: PaypalResolvers;
  CreditCard?: CreditCardResolvers;
  Donation?: DonationResolvers;
  Mutation?: MutationResolvers;
  User?: UserResolvers;
};

export interface MiddlewareMap {
  '*'?: {
    '*'?: gm.Middleware[];
  };
  Paypal?: {
    '*'?: gm.Middleware[];
    id?: gm.Middleware[];
    url?: gm.Middleware[];
  };
  CreditCard?: {
    '*'?: gm.Middleware[];
    id?: gm.Middleware[];
    cardNumber?: gm.Middleware[];
    cardOwner?: gm.Middleware[];
  };
  User?: {
    '*'?: gm.Middleware[];
    paymentOptions?: gm.Middleware[];
  };
  Donation?: {
    '*'?: gm.Middleware[];
    id?: gm.Middleware[];
    sender?: gm.Middleware[];
    recipient?: gm.Middleware[];
    amount?: gm.Middleware[];
  };
  Mutation?: {
    '*'?: gm.Middleware[];
    donate?: gm.Middleware[];
  };
};"
`;
